mkosi/run.py

```diff
@@ -161,6 +161,79 @@ def ensure_exc_info() -> tuple[type[BaseException], BaseException, TracebackType
     return (exctype, exc, tb)
 
 
+@contextlib.contextmanager
+def uncaught_exception_handler(exit: Callable[[int], NoReturn]) -> Iterator[None]:
+    try:
+        yield
+    except SystemExit as e:
+        if ARG_DEBUG.get():
+            sys.excepthook(*ensure_exc_info())
+
+        exit(e.code if isinstance(e.code, int) else 1)
+    except KeyboardInterrupt:
+        if ARG_DEBUG.get():
+            sys.excepthook(*ensure_exc_info())
+        else:
+            logging.error("Interrupted")
+
+        exit(1)
+    except subprocess.CalledProcessError as e:
+        # Failures from qemu, ssh and systemd-nspawn are expected and we won't log stacktraces for those.
+        # Failures from self come from the forks we spawn to build images in a user namespace. We've already done all
+        # the logging for those failures so we don't log stacktraces for those either.
+        if ARG_DEBUG.get() and e.cmd and e.cmd[0] not in ("self", "qemu", "ssh", "systemd-nspawn"):
+            sys.excepthook(*ensure_exc_info())
+
+        # We always log when subprocess.CalledProcessError is raised, so we don't log again here.
+        exit(e.returncode)
+    except BaseException:
+        sys.excepthook(*ensure_exc_info())
+        exit(1)
+    else:
```

Comment:
Is there a deeper reason this is an `else` instead of ` finally`, i.e. doesn't it make sense to flush the streams in the error case as well?

Reply:
So I did it like this because if we call `os._exit()` the finally won't actually run and I didn't really like adding the flushing to every exception block. But we should really always flush, so I reworked this now to always flush.